Erik Bülows blog post (#123)
ropensci/roweb3: content/blog/2021-01-25-coder/index.Rmd

---
slug: "coder"

> title: coder Makes Medical Coding less Messy

Reviewer: Great title!

package_version: 0.13.6
author:
  - Erik Bülow

> date: 2021-01-25

Reviewer: Change the date here and in the folder name

tags:
  - Software Peer Review
  - packages

>   - R

Reviewer: We're taking out the R tag now (very recent change)
```suggestion
```

  - community
  - coder
  - data-munging
# The summary below will be used by e.g. Twitter cards

> description: "A very short summary of your post (~ 100 characters)"

Reviewer: Don't forget a description for the twitter card!

Author: I have now added a short text.

> Many people have realized this. [Charlson](https://pubmed.ncbi.nlm.nih.gov/3558716/), [Elixhauser](https://pubmed.ncbi.nlm.nih.gov/9431328/), [Sloan](https://pubmed.ncbi.nlm.nih.gov/12773842/) and others, have proposed to combine individual codes into broader categories to capture patient comorbidity more holistically. There are still several versions of each index, however. To use those indices in practice is also cumbersome since many data sources (large administrative registers) are big - bigger than what would fit into the random access memory (RAM) of a standard computer.

Reviewer: Perhaps give a vague kudos to the coauthors...

```suggestion
Many people have realized this. [Charlson et al.](https://pubmed.ncbi.nlm.nih.gov/3558716/), [Elixhauser et al.](https://pubmed.ncbi.nlm.nih.gov/9431328/), [Sloan et al.](https://pubmed.ncbi.nlm.nih.gov/12773842/) and others, have proposed to combine individual codes into broader categories to capture patient comorbidity more holistically. There are still several versions of each index, however. To use those indices in practice is also cumbersome since many data sources (large administrative registers) are big - bigger than what would fit into the random access memory (RAM) of a standard computer.
```

> Many people have realized this. [Charlson](https://pubmed.ncbi.nlm.nih.gov/3558716/), [Elixhauser](https://pubmed.ncbi.nlm.nih.gov/9431328/), [Sloan](https://pubmed.ncbi.nlm.nih.gov/12773842/) and others, have proposed to combine individual codes into broader categories to capture patient comorbidity more holistically. There are still several versions of each index, however. To use those indices in practice is also cumbersome since many data sources (large administrative registers) are big - bigger than what would fit into the random access memory (RAM) of a standard computer.

Reviewer: Perhaps a short explanation of what comorbidity is and why it matters?

Author: I have expanded this section a bit.

## R packages

> There are some excellent R-packages to help with both the Charlson and Elixhauser comorbidity indices. [`{comorbidity}`](https://ellessenne.github.io/comorbidity/) by Alessandro Gasparini and [`{icd}`](https://jackwasey.github.io/icd/) by Jack O. Wasey and Michael Lang are very well documented with efficient implementations. To keep up with all different and newly proposed coding versions seems like a daunting, if not impossible task, however. Therefore, the `{coder}` package offers an alternative, more flexible, implementation. It is not hard-wired to Charlson or Elixhauser, although capabilities for those indices are included as well. `{data.table}` is used internally for computational efficiency, while the application programming interface (API) is more inspired by `{tidyverse}`. 

Reviewer: - Remove formatting around package names
- Maybe link to data.table and tidyverse?

## Regular expressions

> In `{coder}`, all relevant codes (from ICD, ATC, NCSP or whatever) are represented in a compact way by their corresponding regular expressions. Peptic ulcer disease for example is recognized as an important comorbidity by the Elixhauser classification. It might be coded by ICD-10 as “K25.x–K28.x” (see [table 2 here](https://journals.lww.com/lww-medicalcare/Abstract/2005/11000/Coding_Algorithms_for_Defining_Comorbidities_in.10.aspx)). This format is well understood by humans; "-" indicates an interval and "x" acts like a wild card for any additional alphanumerical characters. This might be translated into a code list (ignoring the dots): K257, K259, K267, K269, K277, K279, K287, K289. But why those codes only? Why not K250 or K289A? Well, not all codes are used in practice. This list is just an example corresponding to relevant codes from the clinical implementation of ICD-10 (ICD-10-CM) used in 2020 in the USA. The list might look different with data from the Swedish version of ICD-10 (ICD-10-SE) recorded in 2021. Hence, the use of regular expressions will relax the need of a complete and version-specific code list. This also increases the computational speed, which might be crucial if working with large populations (as from a national patient register spanning several decades et cetera). A regular expression for peptic ulcer disease would be `^K2[5-8][79]$` where "^" ("$") marks the beginning (end) of the character string, and where alternative digits are written within brackets (see `?regex` in R for details). 

Reviewer: I'm a bit confused by some of the wording... Where you say "This list is just an example corresponding to relevant codes", do you mean that coder only returns relevant codes? Perhaps add a bit of clarification? Why does it increase computational speed?

Author: I agree! I change "relevant codes ..." to "codes defined in the ..."

Author: I have also add some additional clarification concerning computational efficiency.

By default, there are 413 of those regular expressions included in the `{coder}` package. They cover all conditions recognized by Charlson, Elixhauser, RxRisk V, the comorbidity-polypharmacy score (CPS), as well as for some diagnose specific adverse events classifications after hip or knee replacement. Charlson, Elixhauser and Rx Risk V have 6, 5 and 3 alternative regular expressions for each condition. Peptic ulcer disease for example can also be recognized from 1) `53([1-4][79]0)|V1271`; 2) `53([1-4]([4-69]1|7))`; or 3) `53[1-4][79]` based on three different versions of ICD-9-CM. Additional versions (think of the upcoming ICD-11) are relatively easy to implement by the user using a dedicated `classcodes` object (an enhanced data frame/tibble with alternative regular expressions in each column).

> Although `classcodes` objects are relatively easy to implement, the interpretation of a regular expression might be hard to grasp intuitively. Or what about hypertension according to [Pratt's implementation](https://bmjopen.bmj.com/content/8/4/e021122) of the Rx Risk V classification based on ATC codes: `C0(2(A(B[0-9]{2}|C0[0-5])|DB(0[2-9]|[1-9][0-9]))|3((A[A-Z][0-9]{2}|BA(0[0-9]|1[01]))|DB(01|99)|C([AB][0-9]{2}|C0[01])|EA01)|9(BA0[2-9]|DA0[2-8]|C[A-X][0-9]{2}))`? There are several functions to aid such interpretation: `visualize()` will open the default web browser to show a structural diagram based on the [JavaScript Regular Expression Visualizer](https://jex.im/regulex). `summary()` and `codebook()` combines the features of `{coder}` with the [`{decoder}`](https://cran.r-project.org/package=decoder) package. By so, it is easy to make explainable code lists with both codes and their interpreted meaning.

Reviewer: How is the decoder package relevant? What do summary() and codebook() actually do?

```suggestion
Although `classcodes` objects are relatively easy to implement, the interpretation of a regular expression might be hard to grasp intuitively. Or what about hypertension according to [Pratt's implementation](https://bmjopen.bmj.com/content/8/4/e021122) of the Rx Risk V classification based on ATC codes: `C0(2(A(B[0-9]{2}|C0[0-5])|DB(0[2-9]|[1-9][0-9]))|3((A[A-Z][0-9]{2}|BA(0[0-9]|1[01]))|DB(01|99)|C([AB][0-9]{2}|C0[01])|EA01)|9(BA0[2-9]|DA0[2-8]|C[A-X][0-9]{2}))`? There are several functions to aid such interpretation: `visualize()` will open the default web browser to show a structural diagram based on the [JavaScript Regular Expression Visualizer](https://jex.im/regulex). `summary()` and `codebook()` combine the features of `{coder}` with the [`{decoder}`](https://cran.r-project.org/package=decoder) package. By so, it is easy to make explainable code lists with both codes and their interpreted meaning.
```

Author: I managed to get this "outdated" somehow but I have changed "conbines" to "conbine" in the file and will of couse adress the broader issue as well!

Author: I have added some more explanaitions and examples here.

## Use case

Assume we have a list of patients who received surgery at specified dates:

```{r}
library(coder)
ex_people # Random names and dates included in the package
```

Those patients (among others) were also recorded in a national patient register with date of hospital admissions and corresponding ICD-10 codes. Each patient can have multiple hospital visits over the years. Also, multiple diagnosis might have been recorded at each occasion.

```{r}
ex_icd10
```

Using those two data sets, as well as a classification scheme (a `classcodes` object), we can easily identify all Charlson comorbidities for each patient:

```{r}
ch <- 
  categorize(
    ex_people,                  # patients of interest 
    codedata = ex_icd10,        # Medical codes from national patient register
    cc = "charlson",            # Calculate Charlson comorbidity
    id = "name", code = "icd10" # Specify column names
  )
```

Now, how many of those patients were diagnosed with malignancy?

```{r}
sum(ch$malignancy)
```

> What is the distribution of the combined comorbidity index for each patient?

Reviewer: Consider adding a brief explanation of why this is important in brackets after this statement, just for those of us not in medicine :grin:

e.g., "What is the distribution of the combined comorbidity index for each patient? (This tells us...)

Author: I have added a brief explanation

> ```{r}

Reviewer: Please add some Alt text for people who are visually impaired (see https://blogguide.ropensci.org/technical.html#addfigure for guidance)

barplot(table(ch$charlson))
```

## Further usage

> There are many more examples provided at the [`{coder}`  web page](https://docs.ropensci.org/coder/). It is easy to constrain the time frame to comorbidities recorded during only one year prior to surgery, or to 90 days for adverse events thereafter, or to use different versions of the regular expressions etc. Core functionality is provided by the `categorize()` function used above. Alternatively, more fine grained control is provided through a more explicit chain of commands: `codify() %>% classify() %>% index()`. So far, the package has been used in several research projects (cited by the [JOSS article](https://joss.theoj.org/papers/10.21105/joss.02916)). Contributations are most welcome and I would be very happy to [hear about](https://github.com/ropensci/coder/issues) possible additional use cases. And please note that there are no hard-coded limitations to patients and/or medical data. The package can be used with all types of codified data!

Reviewer: ```suggestion
There are many more examples provided at the [`{coder}`  web page](https://docs.ropensci.org/coder/). It is easy to constrain the time frame to comorbidities recorded during only one year prior to surgery, or to 90 days for adverse events thereafter, or to use different versions of the regular expressions etc. Core functionality is provided by the `categorize()` function used above. Alternatively, more fine grained control is provided through a more explicit chain of commands: `codify() %>% classify() %>% index()`. So far, the package has been used in several research projects (cited by the [JOSS article](https://joss.theoj.org/papers/10.21105/joss.02916)). Contributions are most welcome and I would be very happy to [hear about](https://github.com/ropensci/coder/issues) possible additional use cases. And please note that there are no hard-coded limitations to patients and/or medical data. The package can be used with all types of codified data!
```

## Acknowledgement

> An early version of the `{coder}` package was significantly improved after generous contributions made by [Emely C Zabore](https://github.com/zabore) and [David Robinson](/author/david-robinson/) during the peer-review process for rOpenSci. I therefore wish to express my deepest gratitudes to both of them, as well as to the rOpenSci community at large!

Reviewer: ```suggestion
An early version of the `{coder}` package was significantly improved after generous contributions made by [Emely C Zabore](https://github.com/zabore) and [David Robinson](/author/david-robinson/) during [the peer-review process](https://github.com/openjournals/joss-reviews/issues/2916) for rOpenSci. I therefore wish to express my deepest gratitude to both of them, as well as to the rOpenSci community at large!
```

Author: It seems this suggested change became "outdated" somehow and I can not accept it automatically. I will of course make the corresponding change to the source file, however! I change the link to the rOpenSci review (instead of the JOSS review, in accordance with suggestion above.)
